grpc/transport-auth.go

```diff
@@ -0,0 +1,87 @@
+package grpc
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/credentials"
+)
+
+// transportCredentials is a grpc/credentials.TransportCredentials which supports
+// connecting to, and verifying multiple DNS names
+type transportCredentials struct {
+	configs map[string]*tls.Config
+}
+
+func newTransportCredentials(addrs []string, rootCAs *x509.CertPool, clientCerts []tls.Certificate) (credentials.TransportCredentials, error) {
+	configs := make(map[string]*tls.Config, len(addrs))
+	for _, addr := range addrs {
+		host, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, err
+		}
+		configs[addr] = &tls.Config{
+			ServerName:   host,
+			RootCAs:      rootCAs,
+			Certificates: clientCerts,
+		}
+	}
+	return &transportCredentials{configs}, nil
+}
+
+// ClientHandshake performs the TLS handshake for a client -> server connection
+func (tc *transportCredentials) ClientHandshake(addr string, rawConn net.Conn, timeout time.Duration) (net.Conn, credentials.AuthInfo, error) {
+	ctx := context.Background()
+	var cancel func()
+	if timeout != 0 {
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
+	config, present := tc.configs[addr]
+	if !present {
+		return nil, nil, fmt.Errorf("boulder/grpc: Unexpected name, no TLS configuration present for \"%s\"", addr)
```

Comment:
This was mainly to avoid unnecessary allocations. Since the objects are safe for concurrent use it makes sense to just create them once instead of having to do it each time a new connection is made.